### ISSUE TITLE
"Can't use an undefined value as an ARRAY reference" in dev version 5.985_057

### ISSUE THREAD

**Issue by gwyndaf** [ChordPro/chordpro#191]
## Describe the bug
Error when processing songs with certain instrument configs:
```Can't use an undefined value as an ARRAY reference at /media/nfs/Shared/Music/Technology/ChordPro/git/chordpro/script/../lib/App/Music/ChordPro/Output/PDF/StringDiagrams.pm line 153.```

I've tried to narrow down circumstances and it seems to arise with enharmonic equivalent chords that aren't explicitly defined in instrument config, e.g. song that contains Bb, while my instrument config defines only A# (but not Bb explicitly.

Normally, these get handled nicely, and still work well with release version, so I'm guessing it's just a development glitch, but thought worth flagging.

**Comment by sciurius** [comment]
There have been some changes in this area recently. Please checkout the latest development from git, and if the problem persist, include a small song + config that shows the problem.

**Comment by gwyndaf** [comment]
Thanks. Seems to have gone away with dev version _061, or at least is pre-empted by a new error for _all_ songs:
```
garbage after JSON object, at character offset 7789 (before "\n\n}\n") at /media/nfs/Shared/Music/Technology/ChordPro/git/chordpro/script/../lib/App/Music/ChordPro/Config.pm line 303.
```
Doesn't seem to arise when using `--nouserconfig` so seems like there's stuff in my configs that used to work but now doesn't. I'll have a compare between git `chordpro.json` and my own to see if I can identify what difference might be causing that.

**Comment by gwyndaf** [comment]
Sorry, slight detour via my messy config.

Issue still seems to persist, except line number has shifted:
```
Can't use an undefined value as an ARRAY reference at /media/nfs/Shared/Music/Technology/ChordPro/git/chordpro/script/../lib/App/Music/ChordPro/Output/PDF/StringDiagrams.pm line 160.
```
I'm pretty sure it arises because there's a Bb in the song, but it's not defined in the instrument config. Previously, ChordPro successfully used the A# definition (enharmonic equivalent), which does exist. I can prevent the error arising by either removing Bb from the song, or adding a Bb definition to the config.

However, even if was an invalid chord, I'd expect a different error message, specifically identifying the song and chord.

Here's a sample, of both test song file and instrument config
[Merry_Christmas_Everybody_TEST.cho.txt](https://github.com/ChordPro/chordpro/files/7705626/Merry_Christmas_Everybody_TEST.cho.txt)
[ukulele_c.json.txt](https://github.com/ChordPro/chordpro/files/7705632/ukulele_c.json.txt)

**Comment by sciurius** [comment]
Thanks for the testing files.
After git pull, the unknown chords will be flagged and will not crash the program.

Your config doesn't define `Bb` so it is now a known chord

I'm not sure whether enharmonic equivalents should be substituted. This worked in the past but caused lots of troubles (e.g. a `Bb` chord that comes out as `A#`). What do you think?

**Comment by gwyndaf** [comment]
Great. v _066 gets rid of that error message.

Undefined `Bb` now gets reported as _un_ known (which I think is maybe what you meant), with standard, informative 'unknown chord' messages.

Personally, I've found the enharmonic substitution useful, but only because of my approach to creating and maintaining some instrument configs. I have a single spreadsheet that can export configs for C, G and D tuning ukuleles, with a single set of voicings/fingerings, and just 'rotating' the chord names (using only sharp names). When using flat chords, I discovered that enharmonic substitution helpfully found the corresponding voicing/fingers (as a sharp chord), and gave the required diagram, with flat chord name.

In my own usage, I've never had the wrong chord diagram name returned, e.g. putting `Bb` in a song and getting an `A#` diagram, even if the config defined the chord only as `A#` (unless I transposed in the wrong direction!). But it sounds like there might be cases where that could be an issue. Maybe if there's a `display` name specified in the config it gets more complicated?

If you feel it's more reliable in day-to-day usage to avoid enharmonic substitution, maybe that's best approach. Creating instrument configs is generally a 'one-off' bit of work, and it's not a huge effort just to add corresponding flat chord definitions where I've already got sharp ones: I simply treated it as a time-saving bonus that it seemed to work, but can happily live without it, and just be more rigorous with my instrument definitions!


One other message that's now started appearing: note names like `[g]` are now being reported as unknown chords, even though I've got `"notenames" : true,` defined. I use them as a quick way of writing short melody riffs (not generating diagrams), often within a `{comment: }` so they'll transpose easily. Doesn't seem to affect how they're shown on the page, but might be an unnecessary warning message (unless some usages define and generate a diagram for a single note)?
